### PR TITLE
Fix dataset fingerprinting in DPO/SFT tokenization

### PR DESCRIPTION
**Pull request by qgallouedec**
`DPOTrainer` and `SFTTrainer` tokenize via a nested `tokenize_fn` passed to `Dataset.map` that called `self._tokenize(...)`, capturing the whole trainer. When `datasets` fingerprints the map function it tries to pickle `self`; if that fails it falls back to a **random fingerprint**, silently disabling tokenization caching and logging:

```
Parameter 'function'=<function DPOTrainer._prepare_dataset.<locals>.tokenize_fn ...> couldn't be hashed properly, a random hash was used instead. ...
```

## Background

Not a recent regression: the self-capture dates back to **#3906**; **#5369** only moved it into the `_tokenize` helper. **#6116** made it *visible* — its `evaluate()` override re-tokenizes a raw eval set after the model is accelerate-wrapped, so `self` is no longer picklable and hashing now raises instead of quietly returning a (still-wrong) hash. #6116 also added the `test_evaluate_with_raw_dataset` test that surfaced it.

## Fix

Make `_tokenize` a `@staticmethod` taking the `self`-derived values (`is_vlm`, plus `chat_template` for SFT) explicitly. `tokenize_fn` references it through a local (`tokenize = self._tokenize`, no bound `self`) and gets those values via `fn_kwargs`, like the existing `processing_class`. No `self` in the closure → hashable again.

Scope: **DPO** and **SFT** had the bug — fixed. **Reward** never captured `self` (no bug), but is realigned to the same `_tokenize` static-helper structure for consistency. **GRPO/RLOO** unaffected (tokenize during generation, not via `Dataset.map`).

## Reproducer

Warns on `main`, silent after this PR:

```python
from datasets import load_dataset
from trl import DPOTrainer

dataset = load_dataset("trl-internal-testing/zen", "standard_preference", split="train")
trainer = DPOTrainer(model="trl-internal-testing/tiny-Qwen2ForCausalLM-2.5", train_dataset=dataset)
trainer.evaluate(eval_dataset=dataset)
```

Before the fix:

```console
$ python repro.py
Loading weights: 100%|██████████████████████████████████| 27/27 [00:00<00:00, 1060.31it/s]
Loading weights: 100%|██████████████████████████████████| 27/27 [00:00<00:00, 4642.95it/s]
Parameter 'function'=<function DPOTrainer._prepare_dataset.<locals>.tokenize_fn at 0x7fec260267a0> of the transform datasets.arrow_dataset.Dataset._map_single couldn't be hashed properly, a random hash was used instead. Make sure your transforms and parameters are serializable with pickle or dill for the dataset fingerprinting and caching to work. If you reuse this transform, the caching mechanism will consider it to be different from the previous calls and recompute everything. This warning is only shown once. Subsequent hashing failures won't be shown.
Tokenizing eval dataset: 100%|████████████████████| 17/17 [00:00<00:00, 1208.47 examples/s]
100%|██████████████████████████████████████████████████████| 3/3 [00:00<00:00, 67.07it/s]
```

After the fix:

```
$ python repro.py
Loading weights: 100%|██████████████████████████████████| 27/27 [00:00<00:00, 1369.13it/s]
Loading weights: 100%|██████████████████████████████████| 27/27 [00:00<00:00, 2944.83it/s]
100%|██████████████████████████████████████████████████████| 3/3 [00:00<00:00, 87.99it/s]
```

---

> [!NOTE]
> **Low Risk**
> Refactor of dataset preprocessing only; tokenization behavior is intended to be unchanged aside from enabling correct caching.
> 
> **Overview**
> Fixes **Hugging Face `datasets` fingerprinting** for preference/SFT preprocessing: nested `tokenize_fn` closures no longer capture the trainer instance, so `Dataset.map` can hash the transform and **reuse tokenization caches** instead of falling back to a random fingerprint (and the associated warning).
> 
> **DPO** and **SFT** turn `_tokenize` into a `@staticmethod` and pass trainer-derived flags (`is_vlm`, and `chat_template` for SFT) through `fn_kwargs`. **`tokenize_fn`** calls a local `tokenize = self._tokenize` so the mapped function stays picklable even after Accelerate wraps the model (e.g. raw eval sets re-tokenized in `evaluate()`).
> 
> **RewardTrainer** gets the same static `_tokenize` helper and local binding for consistency; it did not previously close over `self` in the map path.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 1b9ce2da24f01a0d9b28d8177b540e5f048640d3. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: trl/trainer/sft_trainer.py [hunk 1]
```diff
@@ -1558,6 +1581,8 @@ def tokenize_fn(example, processing_class, dataset_text_field, assistant_only_lo
                         "processing_class": processing_class,
                         "dataset_text_field": args.dataset_text_field,
                         "assistant_only_loss": args.assistant_only_loss,
+                        "is_vlm": self._is_vlm,
```
Comment (kashif) [inline trl/trainer/sft_trainer.py:1584]: nice!

**Review by kashif — APPROVED** [review]

File: trl/trainer/dpo_trainer.py [hunk 2]
```diff
@@ -961,36 +965,43 @@ def add_eos(example, eos_token):
             if isinstance(dataset, Dataset):  # `IterableDataset.map` does not support `desc`
                 map_kwargs["desc"] = f"Tokenizing {dataset_name} dataset"
 
-            def tokenize_fn(example, processing_class):
+            # Bind `_tokenize` to a local so `tokenize_fn` doesn't capture `self`: a closure over `self` makes the map
+            # function unhashable, forcing a random fingerprint that silently disables dataset caching.
+            tokenize = self._tokenize
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/trainer/dpo_trainer.py:970]: no this touches only stable trainers

**Comment by albertvillanova** [comment]
@qgallouedec, indeed I had been investigating this issue for several days... :sweat_smile: https://github.com/huggingface/trl/actions/runs/28230236837/job/83631939813
However, I was not fully convinced about this solution proposed by the agent. :thinking: 
The issue only appeared **after** the implementation of `evaluate` #6116. Before, the dataset preparation was called during trainer's init (before calling super.init), and `self` was pickable at that moment.

I will open a PR with the proposal I have been working on.
